collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java

```diff
@@ -130,421 +127,316 @@ public void onCreate(Bundle savedInstanceState) {
+
     public void returnLocation() {
-        Intent i = new Intent();
-        if (setClear || (readOnly && latLng == null)) {
-            i.putExtra(FormEntryActivity.LOCATION_RESULT, "");
-            setResult(RESULT_OK, i);
+        String result = null;
 
-        } else if (isDragged || readOnly || locationFromIntent) {
+        if (setClear || (intentReadOnly && featureId == -1)) {
+            result = "";
+        } else if (isDragged || intentReadOnly || pointFromIntent) {
             Timber.i("IsDragged !!!");
-            i.putExtra(
-                    FormEntryActivity.LOCATION_RESULT,
-                    latLng.latitude + " " + latLng.longitude + " "
-                            + 0 + " " + 0);
-            setResult(RESULT_OK, i);
+            result = formatResult(map.getMarkerPoint(featureId));
         } else if (location != null) {
             Timber.i("IsNotDragged !!!");
+            result = formatResult(location);
+        }
 
-            i.putExtra(
-                    FormEntryActivity.LOCATION_RESULT,
-                    getResultString(location)
-            );
-            setResult(RESULT_OK, i);
+        if (result != null) {
+            setResult(RESULT_OK, new Intent().putExtra(FormEntryActivity.LOCATION_RESULT, result));
         }
         finish();
     }
 
-    public String getResultString(Location location) {
-        return String.format("%s %s %s %s", location.getLatitude(), location.getLongitude(), location.getAltitude(), location.getAccuracy());
-    }
-
-    private String truncateFloat(float f) {
-        return new DecimalFormat("#.##").format(f);
-    }
-
     @SuppressLint("MissingPermission") // Permission handled in Constructor
-    private void setupMap(GoogleMap googleMap) {
-        map = googleMap;
-        if (map == null) {
-            ToastUtils.showShortToast(R.string.google_play_services_error_occured);
+    public void initMap(MapFragment newMapFragment) {
+        if (newMapFragment == null) {
             finish();
             return;
         }
-        helper = new MapHelper(this, map, selectedLayer);
-        map.setMyLocationEnabled(true);
-        map.getUiSettings().setCompassEnabled(true);
-        map.getUiSettings().setMyLocationButtonEnabled(false);
-        map.getUiSettings().setZoomControlsEnabled(false);
 
-        markerOptions = new MarkerOptions();
-        helper = new MapHelper(this, map, selectedLayer);
+        map = newMapFragment;
+        map.setDragEndListener(this::onDragEnd);
+        map.setLongPressListener(this::onLongPress);
 
-        ImageButton acceptLocation = findViewById(R.id.accept_location);
+        if (map instanceof GoogleMapFragment) {
+            helper = new MapHelper(this, ((GoogleMapFragment) map).getGoogleMap(), selectedLayer);
+        } else if (map instanceof OsmMapFragment) {
+            helper = new MapHelper(this, ((OsmMapFragment) map).getMapView(), this, selectedLayer);
+        }
```

Comment:
Is it okay for `helper` to be unassigned? Maybe add a final `else` that throws?

Reply:
Done.

This `if` statement will soon be refactored out of existence (the proper place for the MapHelper logic is in the GoogleMapFragment or OsmMapFragment; this is a temporary situation until MapHelper is refactored to fit the new MapFragment interface).